herald/herald/Sensor/BLE/BLEReceiver.swift

```diff
@@ -895,28 +900,26 @@ class ConcreteBLEReceiver: NSObject, BLEReceiver, BLEDatabaseDelegate, CBCentral
             case BLESensorConfiguration.payloadCharacteristicUUID:
                 device.payloadCharacteristic = characteristic
                 logger.debug("didDiscoverCharacteristicsFor, found payload characteristic (device=\(device))")
-            case BLESensorConfiguration.legacyPayloadCharacteristicUUID:
-                if characteristics.count == 1 {
-                    device.legacyPayloadCharacteristic = characteristic
-                    logger.debug("didDiscoverCharacteristicsFor, found legacy payload characteristic (device=\(device))")
-                } else {
-                    logger.debug("didDiscoverCharacteristicsFor, found legacy payload characteristic but discarding as there are more characteristics, assuming new ble (device=\(device))")
-                }
+            case BLESensorConfiguration.interopOpenTracePayloadCharacteristicUUID:
+                device.legacyPayloadCharacteristic = characteristic
+                logger.debug("didDiscoverCharacteristicsFor, found legacy payload characteristic (device=\(device))")
             default:
                 logger.fault("didDiscoverCharacteristicsFor, found unknown characteristic (device=\(device),characteristic=\(characteristic.uuid))")
             }
         }
-        // Android -> Read payload
-        if device.operatingSystem == .android {
-            var payloadCharacteristic = device.payloadCharacteristic
-            if payloadCharacteristic == nil {
-                payloadCharacteristic = device.legacyPayloadCharacteristic
-                if nil != device.legacyPayloadCharacteristic {
-                    writeLegacyPayload("didDiscoverCharacteristicsFor|android", peripheral: peripheral)
-                }
+        // OpenTrace -> Read payload -> Write payload -> Disconnect
+        if let legacyPayloadCharacteristic = device.legacyPayloadCharacteristic, device.signalCharacteristic == nil {
+            if device.payloadData == nil || device.timeIntervalSinceLastPayloadDataUpdate > BLESensorConfiguration.payloadDataUpdateTimeInterval {
```

Comment:
What if device supports transmitting multiple payloads?

Reply:
Service discovery is only searching for Herald, OpenTrace, and Advert based protocol.
- Advert based protocol will not trigger connection
- If Herald (has signal characteristic) and OpenTrace (has legacy payload characteristic) are both found o the same target, Herald takes priority, OpenTrace is ignored, because this only happens when the target is actual Herald with OpenTrace support enabled.
- If OpenTrace is found without Herald (no signal characteristic), this is a genuine OpenTrace only device, thus the above code path is executed to read from OpenTrace, write to OpenTrace, before disconnection.